src/DataObject/DataObjectBatchProcessor.php

```diff
@@ -54,13 +56,30 @@ public function removeDocuments(string $indexSuffix, array $documents): array
                     continue;
                 }
 
-                // set up the separate job to update these dependencies
+                // set up the separate index job to update these dependencies
                 $job = IndexJob::create($indexSuffix, $dependencies, Indexer::METHOD_ADD, null, false);
                 $this->run($job);
 
                 continue;
             }
 
+            // Allow extensions to handle dependency resolution for versioned objects
+            // This is useful for modules that need special handling (e.g., Fluent)
+            // Extensions can set $handled to true and optionally provide $dependencies to process
+            $handled = false;
+            $dependencies = [];
+            $this->invokeWithExtensions('updateRemoveVersionedDocument', $indexSuffix, $doc, $handled, $dependencies);
```

Comment:
see if we can combine this with the unversioned code above as it is very similar. Could even be a feature flag/config like `use_synchronous_dependencies`. I'd be happy to deprecate the `RemoveDataObjectJob` as part of this as I'd rather move to a system that supports more than versioned objects by default.